core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2416,14 +2419,56 @@ class KafkaApis(val requestChannel: RequestChannel,
         numAppends.decrementAndGet()
         skippedMarkers += 1
       } else {
-        val controlRecords = partitionsWithCompatibleMessageFormat.map { partition =>
-          val controlRecordType = marker.transactionResult match {
-            case TransactionResult.COMMIT => ControlRecordType.COMMIT
-            case TransactionResult.ABORT => ControlRecordType.ABORT
+        val controlRecordType = marker.transactionResult match {
+          case TransactionResult.COMMIT => ControlRecordType.COMMIT
+          case TransactionResult.ABORT => ControlRecordType.ABORT
+        }
+
+        val markerResults = new ConcurrentHashMap[TopicPartition, Errors]()
```

Comment:
to understand: 
 * markerResults represents the to-be appended error result for each topic partition in the marker. 
 * it is concurrent since it also gets updated via appendRecords() response callback.
 * it is updated separately in the new group coordinator as we append through the new runtime framework and not via appendRecords().
 * we check `if (partitionsWithCompatibleMessageFormat.size == markerResults.size)` to only call `maybeSendResponseCallback` once for all partitions in a marker. `numAppends` serves a similar purpose but it applies to a one final response for all markers.

does this sound right?

Reply:
yep, it sounds right.